### PR TITLE
Added implementation-specific version filtering

### PR DESCRIPTION
**Pull request by sifraser**

### CODE REVIEW

File: azuki-core/build.gradle [hunk 1]
```diff
@@ -6,6 +6,7 @@ plugins {
 dependencies {
     implementation 'org.jetbrains.kotlin:kotlin-test-junit'
     implementation 'com.fasterxml.jackson.module:jackson-module-kotlin:2.12.5'
+    implementation 'com.vdurmont:semver4j:3.1.0'
```
Comment (donbex) [inline azuki-core/build.gradle]: ⚠️ This package is no longer maintained, but there is an active fork: https://github.com/semver4j/semver4j

File: azuki-core/src/main/kotlin/com/anaplan/engineering/azuki/core/system/Implementation.kt [hunk 2]
```diff
@@ -16,12 +16,29 @@ interface Implementation<AF : ActionFactory, CF : CheckFactory, QF : QueryFactor
...
         fun <AF : ActionFactory, CF : CheckFactory, QF : QueryFactory, AGF : ActionGeneratorFactory> locateImplementations(): List<Implementation<AF, CF, QF, AGF, *>> {
             val loader = ServiceLoader.load(Implementation::class.java)
             return loader.iterator().asSequence().filterIsInstance<Implementation<AF, CF, QF, AGF, *>>().toList()
         }
     }
+
+    interface VersionFilter {
+
+        /**
+         * Provides a means for an implementation to skip verification of scenario based on version it was introduced
```
Comment (donbex) [inline azuki-core/src/main/kotlin/com/anaplan/engineering/azuki/core/system/Implementation.kt:31]: ⛏️ I find this sentence a bit hard to parse. I had to read it three times to be sure of what it meant.

File: azuki-core/src/test/kotlin/com/anaplan/engineering/azuki/core/system/SemanticVersionFilterTest.kt [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+import kotlin.test.Test
+import kotlin.test.assertEquals
+
+class SemanticVersionFilterTest {
+
+    @Test
+    fun noScenarioVersion() {
+        val semanticVersionFilter = SemanticVersionFilter("1.0.0")
+        assertEquals(true, semanticVersionFilter.canVerify(null))
+        assertEquals(true, semanticVersionFilter.canVerify(null))
```
Comment (donbex) [inline azuki-core/src/test/kotlin/com/anaplan/engineering/azuki/core/system/SemanticVersionFilterTest.kt]: ⛏️ Why assert the same thing twice?

File: azuki-core/src/main/kotlin/com/anaplan/engineering/azuki/core/runner/JUnitScenarioRunner.kt [hunk 4]
```diff
@@ -199,6 +200,16 @@ class JUnitScenarioRunner<
                         } else {
                             testClass.primaryConstructor!!.call(*parameters)
                         }
+                        val sinceAnnotation = build.getAnnotation(Since::class.java)
+                        val scenarioVersion =
+                            sinceAnnotation?.implementationVersion?.filter { it.name == implementationInstance.implementationName }
+                                ?.singleOrNull()
+                        if (implementationInstance.supportsScenarioVersion(scenario, scenarioVersion) == false) {
```
Comment (donbex) [inline azuki-core/src/main/kotlin/com/anaplan/engineering/azuki/core/runner/JUnitScenarioRunner.kt:207]: ⛏️ I know you're keen on always using positive conditionals, but I find this harder to read than `if (!implementationInstance.supportsScenarioVersion(scenario, scenarioVersion))`.

**Review by donbex — CHANGES_REQUESTED** [review]

File: azuki-core/build.gradle (hunk 1 above)
Comment (sifraser) [reply azuki-core/build.gradle]: I don't understand why that would be an issue, but :shrug:

File: azuki-core/src/main/kotlin/com/anaplan/engineering/azuki/core/runner/JUnitScenarioRunner.kt (hunk 4 above)
Comment (sifraser) [reply azuki-core/src/main/kotlin/com/anaplan/engineering/azuki/core/runner/JUnitScenarioRunner.kt:207]: But that doesn't mean the same thing!

File: azuki-core/src/test/kotlin/com/anaplan/engineering/azuki/core/system/SemanticVersionFilterTest.kt (hunk 3 above)
Comment (sifraser) [reply azuki-core/src/test/kotlin/com/anaplan/engineering/azuki/core/system/SemanticVersionFilterTest.kt]: There were two args originally, auto-refactor can't catch

File: azuki-core/src/main/kotlin/com/anaplan/engineering/azuki/core/runner/JUnitScenarioRunner.kt (hunk 4 above)
Comment (donbex) [reply azuki-core/src/main/kotlin/com/anaplan/engineering/azuki/core/runner/JUnitScenarioRunner.kt:207]: Ah, I missed that `implementationInstance.supportsScenarioVersion(scenario, scenarioVersion)` is nullable.

**Review by donbex — APPROVED** [review]
